File: patterns/stack/monotonic_stack/first_greater.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

vector<int> find_greaters(const vector<int> &nums)
{
    // Stage 1: initialization
    stack<int> st;
    vector<int> result(nums.size(), -1);

    for (int i = nums.size() - 1; i >= 0; i--)
    {
        while (!st.empty() && nums[i] >= st.top())
            st.pop();

        if (!st.empty())
            result[i] = st.top();

        st.push(nums[i]);
    }

    return result;
}

int main()
{
    vector<int> nums = {2, 5, 1, 3, 0, 4};
    cout << "Numbers: " << endl;
    for (int num : nums)
    {
        cout << num << " ";
    }

    vector<int> result = find_greaters(nums);
    cout << "Result: " << endl;
    for (int num : result)
    {
        cout << num << " ";
    }

    return 0;
}

File: patterns/stack/monotonic_stack/first_greater.py
"""
Given a list of numbers nums, we should return a new list where
the index of original contains the first element greater than its value.
Otherwise return -1.

e.g.
nums = [2,5,1,3,0,4]
output = [5,-1,3,4,4,-1]
"""


def find_greaters(nums: list[int]) -> list[int]:
    result: list[int] = [-1] * len(nums)
    stack: list[int] = []

    i: int = len(result) - 1
    while i >= 0:
        if len(stack) == 0:
            stack.append(nums[i])
            i -= 1
        elif nums[i] < stack[-1]:
            result[i] = stack[-1]
            stack.append(nums[i])
            i -= 1
        else:
            stack.pop()

    return result


def find_greaters_optimized(nums: list[int]) -> list[int]:
    n = len(nums)
    stack = []
    result = [-1] * n

    for i in range(n - 1, -1, -1):
        while stack and nums[i] >= stack[-1]:
            stack.pop()

        if stack:
            result[i] = stack[-1]

        stack.append(nums[i])

    return result


if __name__ == "__main__":
    nums: list[int] = [2, 5, 1, 3, 0, 4]
    print(f"Numbers: {nums}")
    result: list[int] = find_greaters(nums)
    result_optimized: list[int] = find_greaters_optimized(nums)
    print(f"Result: {result}")
    print(f"Result optimized: {result_optimized}")
